### ISSUE TITLE
Hyprland with on-click causes weird behavior

### ISSUE THREAD

**Issue by SeanMcnabb** [Alexays/Waybar#1850]
I have a custom module that launches rofi when clicked. In sway it works normally but causes some weird problems in Hyprland. After clicking on the module, it will show rofi normally, but if you click anywhere on waybar after that, it will act as if you clicked the module again and either show rofi again, or attempt to. It will stop if you right click the waybar before clicking it again. It doesn't happen with every application. For instance, I was able to get this to happen when launching rofi or kwrite, but launching wofi or alacritty, the behavior is as expected. 

My workaround for now is to launch rofi with gamemoderun which prevents this from happening.

**Comment by SeanMcnabb** [comment]
The following PR fixes this issue: https://github.com/Alexays/Waybar/pull/1784

**Comment by SeanMcnabb** [comment]
So after some further testing, the above PR is only a partial fix. It will prevent the application from launching another instance after clicking anywhere on the waybar, but the waybar won't register any input until you click on it a second time. So after clicking a button with "on-click" and launching rofi or some other programs, it will then require you to click anywhere on the waybar or else it won't accept any input such as scrolling or other clicks

**Comment by justinsandhu** [comment]
Im seeing the same behavior with an on-click script that is running a swaymsg command. @SeanMcnabb  what are you trying to run?

**Comment by Narice** [comment]
I'm having the exact same issue @SeanMcnabb!
@justinsandhu, it basically happens with everything, every tray icon action is broken, and same thing with on-click events on modules (for example launching pavucontrol)
Should we ping someone for this issue?

**Comment by rubyowo** [comment]
I'm facing this problem aswell. Clicking on any module with `on-click` will make its action repeat if you  click again, the only way i can stop it is to right click on any tray icon.

**Comment by Narice** [comment]
Yes, if you right rick each time before left clicking, it works, should we ping alexays?
I don't know where that would come from since it affects the whole bar on Hyprland
@justinsandhu were you on sway? if that's the case, do you still encounter the issue?

**Comment by k1gen** [comment]
I have the same weird behavior on hyprland-git and waybar-hyprland-git on Arch

**Comment by Narice** [comment]
Oh, so that's a problem on all wlroots' based compositors maybe?
In that case @SeanMcnabb, could you update the title to reflect this?

**Comment by SeanMcnabb** [comment]
@Narice I have tested this in sway and do not experience the same behavior. This is happening only on hyprland

**Comment by mrbman** [comment]
I have the same behavior. Tried on both waybar-hyprland-git and waybar-hyprland on Arch

**Comment by rubyowo** [comment]
@Narice Just confirmed. This issue is hyprland only.

**Comment by Narice** [comment]
@rubyowo thanks! that's really useful, I'll try some stuff to see if the issue is coming from waybar or Hyprland.

**Comment by riley-martin** [comment]
I am also using `hyprland-git` and `waybar-hyprland-git` and have this problem as well. Also it seems like tooltips stop working after I have clicked anywhere in the bar; I don't know if that is related to this problem or not.

**Comment by abhakta-47** [comment]
I looked into the issue. It looks like `Task::handle_button_release(GdkEventButton *bt)` in `src/modules/wlr/taskbar.cpp` is not immediately called after mouse button is released in Hyprland. So, it acts like mouse button is still clicked. Not sure if the underlying issue is with hyprland or waybar.

**Comment by mrbman** [comment]
I tested with EWW's onclick function (not running waybar) on Hyprland and I get the same issue, where once I've clicked anything on the EWW bar the same action happens for every subsequent click.

Also tested both EWW and Waybar on Sway and both work fine, and do not show this issue, thus seems to be an issue with Hyperland. Running _hyprland-git_  and _waybar-hyprland-git_ FYI.

**Comment by Rydgel** [comment]
Weird I don't have the issue, `hyprland-git` and `waybar-hyprland-git`. The on-click actions I have is muting/unmuting microphone, muting/unmuting speaker and switching workspaces. Systray and hover popups works fine as well.

**Comment by Aleksanaa** [comment]
> Weird I don't have the issue, `hyprland-git` and `waybar-hyprland-git`. The on-click actions I have is muting/unmuting microphone, muting/unmuting speaker and switching workspaces. Systray and hover popups works fine as well.

That of course doesn't cause the problem. Try setting "on-click" to launch a gui program instead.

**Comment by Stetsed** [comment]
Experiencing this issue aswell on hyprland-git and waybar-hyprland-git. The onclick actions launch rofi menus but then it keeps launching that rofi menu when trying to do anything

**Comment by iMaciaveller** [comment]
I am also using hyprland-git and waybar-hyprland-git and have this problem as well.

**Comment by Aleksanaa** [comment]
> I am also using hyprland-git and waybar-hyprland-git and have this problem as well.

Hey, please refer this issue https://github.com/hyprwm/Hyprland/issues/1348. The developers seems to have no idea how to solve the issue.

**Comment by PROxZIMA** [comment]
> My workaround for now is to launch rofi with gamemoderun which prevents this from happening.

The launch delay is just HUGE. Any recommended settings to launch rofi/tofi quickly using gamemoderun?

**Comment by SeanMcnabb** [comment]
> > My workaround for now is to launch rofi with gamemoderun which prevents this from happening.
> 
> The launch delay is just HUGE. Any recommended settings to launch rofi/tofi quickly using gamemoderun?

For me the delay is only like .25 seconds or something, it's pretty low. I'm not sure what the difference would be.

**Comment by ShabirK21** [comment]
> I have a custom module that launches rofi when clicked. In sway it works normally but causes some weird problems in Hyprland. After clicking on the module, it will show rofi normally, but if you click anywhere on waybar after that, it will act as if you clicked the module again and either show rofi again, or attempt to. It will stop if you right click the waybar before clicking it again. It doesn't happen with every application. For instance, I was able to get this to happen when launching rofi or kwrite, but launching wofi or alacritty, the behavior is as expected.
> 
> My workaround for now is to launch rofi with gamemoderun which prevents this from happening.

does gamemoderun still work for you? it worked previously for me but doesnt now

**Comment by Stetsed** [comment]
Still experiencing this issue with "waybar-hyprland-git" and "hyprland"

**Comment by osalas88** [comment]
In case some one here is not informed.

The current work around appears to be adding some delay to the launching of the application. [See this thread ](https://github.com/hyprwm/Hyprland/issues/1348#issuecomment-1565604984)

Simply adding:

`sleep 0.1 && ...`

before calling your application should fix the issue.

**Comment by aryklein** [comment]
> In case some one here is not informed.
> 
> The current work around appears to be adding some delay to the launching of the application. [See this thread ](https://github.com/hyprwm/Hyprland/issues/1348#issuecomment-1565604984)
> 
> Simply adding:
> 
> `sleep 0.1 && ...`
> 
> before calling your application should fix the issue.

Thanks, that helped me

**Comment by cjuniorfox** [comment]
Same here. Fedora 38 with [solopasha/hyprland](https://copr.fedorainfracloud.org/coprs/solopasha/hyprland/) Copr. Adding "gamemoderun" or some delay works around the issue

**Comment by ipg0** [comment]
@osalas88 thanks a lot, hepled me.

**Comment by LoneWolf4713** [comment]
Hi this bug still persists as of Feb 2024, and no 0.1 delay doesn't work for me. clicking on waybar after opening whatever dialog works

**Comment by stevekanger** [comment]
Is there any more news on this? I am having the same issue. I'm on sway not hyprland. I switched to waybar after just using the sways built in bar for a ever. I made some custom launchers and if I click on custom launcher and then click anywhere on waybar the same click action keeps happening. The only way to get it to stop is to then right click anywhere on waybar. I can get it to stop if I add sleep 0.3 before the command. Not an ideal solution.

**Comment by justinsandhu** [comment]
> @Narice Just confirmed. This issue is hyprland only.

how did you confirm this?  This issue is present in sway as well.

**Comment by PinheiroCosta** [comment]
Originally posted by: @zenmaster3000 
> In case some one here is not informed.
> 
> The current work around appears to be adding some delay to the launching of the application. [See this thread ](https://github.com/hyprwm/Hyprland/issues/1348#issuecomment-1565604984)
> 
> Simply adding:
> 
> `sleep 0.1 && ...`
> 
> before calling your application should fix the issue.

The sleep command works as a workaround for a race condition that occurs when the application is launched immediately on click. Without the delay, the process may start before the environment is ready — for example, before the compositor finishes handling the click event or before the focus context stabilizes.

By adding a small delay (e.g., sleep 0.1 && ...), we give the system enough time to settle, ensuring that the command executes reliably.

In short, the issue is not with the application itself, but with the timing of the event dispatch within Waybar or the display server. The sleep mitigates this by preventing the launch from happening too early.

https://github.com/Alexays/Waybar/issues/4608

**Comment by Alexays** [comment]
Closing — this was resolved/answered in the thread (environment, upstream, or a library limitation — not a Waybar bug still open). Feel free to reopen if you're still stuck. 🙏
